src/shared/virtualFilesystem.ts

```diff
@@ -0,0 +1,115 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+
+type StatNoType = Omit<vscode.FileStat, 'type'>
+
+/**
+ * Basic 'provider' for a file.
+ *
+ * Memory-managment is left up to the caller.
+ * Contents for the file are read as-needed based on the `onDidChange` event.
+ */
+export interface FileProvider {
+    onDidChange: vscode.Event<void>
+    read(): Promise<Uint8Array> | Uint8Array
+    stat(): Promise<StatNoType> | StatNoType
+    write(content: Uint8Array, options: { create: boolean; overwrite: boolean }): Promise<void> | void
+}
+
+/**
+ * Bare-bones file system to support in-memory operations on single documents.
+ * Does not support directories.
+ */
+export class VirualFileSystem implements vscode.FileSystemProvider {
```

Comment:
Extremely minimal implementation that delegates on a per-URI basis.  `dynamicResources` could use this as well.